resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeHelper.scala

```diff
@@ -0,0 +1,180 @@
+        "memory-mb"
+      } else k, v)
+    }
+
+    logDebug(s"Custom resource types: $resourceTypes")
+    resourceTypes.foreach { rt =>
+      val resourceName: String = rt._1
+      val (amount, unit) = getAmountAndUnit(rt._2)
+      logDebug(s"Registering resource with name: $resourceName, amount: $amount, unit: $unit")
+
+      try {
+        val resInfoClass = Utils.classForName(
+          "org.apache.hadoop.yarn.api.records.ResourceInformation")
+        val setResourceInformationMethod =
+          resource.getClass.getMethod("setResourceInformation", classOf[String],
+            resInfoClass)
+
+        val resourceInformation =
+          createResourceInformation(resourceName, amount, unit, resInfoClass)
+        setResourceInformationMethod.invoke(resource, resourceName, resourceInformation)
+      } catch {
+        case e: InvocationTargetException =>
+          if (e.getCause != null) {
+            throw e.getCause
+          } else {
+            throw e
+          }
+        case NonFatal(e) =>
+          logWarning(resourceTypesNotAvailableErrorMessage, e)
+      }
+    }
+    resource
+  }
+
+  def getCustomResourcesAsStrings(resource: Resource): String = {
+    if (resource == null) {
+      throw new IllegalArgumentException("Resource parameter should not be null!")
+    }
+
+    if (!ResourceTypeHelper.isYarnResourceTypesAvailable()) {
+      logWarning(resourceTypesNotAvailableErrorMessage)
+      return ""
+    }
+
+    var res: String = ""
+    try {
+      val resUtilsClass = Utils.classForName(
+        "org.apache.hadoop.yarn.util.resource.ResourceUtils")
+      val getNumberOfResourceTypesMethod = resUtilsClass.getMethod("getNumberOfKnownResourceTypes")
+      val numberOfResourceTypes: Int = getNumberOfResourceTypesMethod.invoke(null).asInstanceOf[Int]
+      val resourceClass = Utils.classForName(
+        "org.apache.hadoop.yarn.api.records.Resource")
+
+      // skip memory and vcores (index 0 and 1)
+      for (i <- 2 until numberOfResourceTypes) {
+        val getResourceInfoMethod = resourceClass.getMethod("getResourceInformation",
+          classOf[Int])
+        res ++= getResourceInfoMethod.invoke(resource, i.asInstanceOf[AnyRef]).toString()
+      }
+    } catch {
+      case e: InvocationTargetException =>
+        if (e.getCause != null) {
+          throw e.getCause
+        } else {
+          throw e
+        }
+      case NonFatal(e) =>
+        logWarning(resourceTypesNotAvailableErrorMessage, e)
+
+    }
+    res
+  }
+
+  def getAmountAndUnit(s: String): (Long, String) = {
+    try {
+      val AMOUNT_AND_UNIT_REGEX(amount, unit) = s
+      (amount.toLong, convertToYarnResourceFormat(unit))
+    } catch {
+      case _: MatchError => throw new IllegalArgumentException(
+        s"Value of resource type should match pattern $AMOUNT_AND_UNIT_REGEX, unmatched value: $s")
+    }
+  }
+
+  def convertToYarnResourceFormat(unit: String): String = {
```

Comment:
According to the Resource Model documentation page of YARN, for every new custom resource type, the specified units in the table are available.
I see your concern, but I don't know how to resolve this situation as Spark uses m and M for Mega unit. How would I differentiate this from micro? 
Suppose someone defines a new resource type (countable) that could have units of mega and giga, and someone requests 2g of that resource.
If the unit is used as is, YARN cannot understand it as g is not a valid unit.

Reply:
Well, I suppose the only real problem here is "m" and "M", since "g" and "G" can both unambiguously mean giga.

To be honest, as far as I know, spark uses "m" only for its memory parameters. YARN requires that resource types have their unit specified in advance (in resource-types.xml, I think), so it will throw an exception if someone specifies "m" but the resource accepts "M".

I still recommend removing this function because it can silently cause wrong behavior, and fixing it later would be hard because we would have to break backwards compatibility (all jobs that used "m" when they really should have used "M" will have to be migrated).

Reply:
Yes, with g/G there is no issue. 
I see Spark uses 'm' only for memory nowadays.
When you define a custom resource type with YARN, you can indeed specify its unit in resource-types.xml, meaning this would be the default unit.
So back to your example, if a resource has M as defined default unit with YARN's config, it does also mean that YARN would accept 'm' as a unit and it will convert the amount of 'm' to 'M' (millis to mega).
More precisely, default units are for cases when you request a resource but don't specify the units, e.g. request 5 and YARN will understand it as request 5M.
So in this sense, requesting any amount of 'm' of a custom resource is not an exceptional case in my opinion.
On the other hand I understand your concern and I see that this conversion is hacky and not clean nor future proof.
I can come up with the following: 
- 'g' (giga) will auto-convert to 'G'
- 'm' / 'M' (millis / mega) will not be auto-converted as it ambiguous (milli vs mega), in this case Spark documentation should be clear that if 'm' is specified then millis will be requested, if 'M' is specified then mega will be requested from YARN.
- 'p' / 'P' (pico / peta): same as millis/mega

Reply:
> So back to your example, if a resource has M as defined default unit with YARN's config, it does also mean that YARN would accept 'm' as a unit and it will convert the amount of 'm' to 'M' (millis to mega).

YARN does this? I might have missed something.

I like your proposal. Since you now disallow requesting memory via this interface (great idea in my opinion), we don't have to worry about the weird case where someone requests 500 megabytes of memory with "500m". To be honest, I would not autoconvert 'g' to 'G', but I do not feel strongly about it.